### PR TITLE
Refactoring widgets

### PR DESCRIPTION
**Pull request by lqez**
I'd refactored widgets simpler and added tests for custom attributes.

### CODE REVIEW

File: django_summernote/settings.py [hunk 1]
```diff
@@ -38,76 +38,69 @@ def get_attachment_model():
...
 SETTINGS_USER = getattr(settings, 'SUMMERNOTE_CONFIG', {})
 SETTINGS_DEFAULT = {
+    # Using SummernoteWidget(iframe widget) for admin pages by default
     'iframe': True,
+
+    # These strings will be assumed as empty.
     'empty': ('<p><br/></p>', '<p><br></p>'),
 
-    'width': 720,
-    'height': 480,
```
[the remark this answers is not in the document]
Comment (jaddison) [reply django_summernote/settings.py]: To be clear the trouble was in my project's admin area, where it uses the iframe widget.

File: django_summernote/templates/django_summernote/widget_common.html [hunk 2]
```diff
@@ -143,7 +143,7 @@
 {% endif %}
     });
 }
-{% if not lazy %}
+{% if not config.lazy %}
```
Comment (jaddison) [inline django_summernote/templates/django_summernote/widget_common.html]: With `lazy` set to `True`, I believe this line breaks iframe widgets

Comment (jaddison) [reply django_summernote/templates/django_summernote/widget_common.html]: Before this change, loading the iframe didn't have access to to the `lazy` param, so it properly always called the `initSummernote_<id>` function. Now that it always has access to it, `lazy` set to `True` breaks this.

Comment (jaddison) [reply django_summernote/templates/django_summernote/widget_common.html]: I'm guessing we want something including the `iframe` value, like:
```
{% if not iframe and not config.lazy %}
```
```

File: README.md [hunk 3]
```diff
@@ -157,24 +152,8 @@ In settings.py,
...
-            os.path.join(STATIC_URL, 'django_summernote/django_summernote.css'),
-        ),
-        'default_js': (
-            os.path.join(STATIC_URL, 'django_summernote/jquery.ui.widget.js'),
-            os.path.join(STATIC_URL, 'django_summernote/jquery.iframe-transport.js'),
-            os.path.join(STATIC_URL, 'django_summernote/jquery.fileupload.js'),
-            os.path.join(STATIC_URL, 'django_summernote/summernote.min.js'),
-        ),
+        # You can disable attachment feature.
+        'disable_attachment': False,
```
Comment (jaddison) [inline README.md]: The issue I have with this is that you still need to put in two or more settings to disable uploads properly `'disable_attachment': True` and `'summernote': { 'disableDragAndDrop': True }` as well as remove the toolbar button appropriately. 

Is this by design, to allow for inlining image uploads as `data` values in the widget output? If so, then ignore my comment! 🤔💥

See my comment in my pull at https://github.com/summernote/django-summernote/pull/279#issuecomment-383211026

**Comment by lqez** [comment]
@jaddison At first, thank you for your review. I'll follow up line by line.

File: README.md (hunk 3 above)
Comment (lqez) [reply README.md]: I agree that these options are pretty easy to be misunderstood. Yes, it caused by Summernote design. 

Summernote is designed to attach images as base64-encoded, but I think most of Django users need file-based attachment system like how we've done before. So I added `Attachment` model in `django-summernote` and also added a way to customize image uploading in `summernote`.

Thus, `disableDragAndDrop` only disables drag-and-drop feature of `Summernote` and `disable_attachment` only disable `attachment` feature of `django-summernote`.

If we want to add an option for disabling the whole features of image uploading, it must be `disable_image` or something else, not `disable_attachment`.

File: django_summernote/settings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lqez) [reply django_summernote/settings.py]: This was my fault. I have to add a test for this case. Thanks.

File: README.md (hunk 3 above)
Comment (jaddison) [reply README.md]: Understood, that history lesson makes sense - thanks for taking the time to explain it. I think what you've got makes sense then.

File: django_summernote/templates/django_summernote/widget_common.html [hunk 4]
```diff
@@ -73,7 +72,7 @@
                     onBlur: function() {
                         origin.value = $sn.summernote('code');
                     },
-                    {% if not disable_upload %}
+                    {% if not config.disable_attachment %}
```
Comment (jaddison) [inline django_summernote/templates/django_summernote/widget_common.html]: Not sure if you were done your changes, but I wanted to mention again that like the `config.lazy` option, `config.disable_attachment` here won't be accurate in an iframe scenario as per my comment at https://github.com/summernote/django-summernote/pull/282#discussion_r184469183

File: django_summernote/settings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jaddison) [reply django_summernote/settings.py]: As a further clarification on this, I did not have any default width/height set in my `SUMMERNOTE_CONFIG` Django settings.

**Comment by jaddison** [comment]
I wasn't sure if you were done your PR changes/updates, so I apologize if I'm being repetitive! 😇

File: django_summernote/settings.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (lqez) [reply django_summernote/settings.py]: Thanks. Now it supports old-style settings via `_copy_old_configs` function.

File: django_summernote/templates/django_summernote/widget_common.html (hunk 4 above)
Comment (jaddison) [reply django_summernote/templates/django_summernote/widget_common.html]: @lqez sorry for the late comment - but I think this is still a problem (and I notice it's still in `master` at https://github.com/summernote/django-summernote/blob/master/django_summernote/templates/django_summernote/widget_common.html#L75)

**Pull request merged**